pkg/clusteragent/custommetrics/provider.go

```diff
@@ -34,24 +36,101 @@ type datadogProvider struct {
 }
 
 // NewDatadogProvider creates a Custom Metrics and External Metrics Provider.
-func NewDatadogProvider(client dynamic.Interface, mapper apimeta.RESTMapper, store Store) provider.MetricsProvider {
+func NewDatadogProvider(ctx context.Context, client dynamic.Interface, mapper apimeta.RESTMapper, store Store) provider.MetricsProvider {
 	maxAge := config.Datadog.GetInt64("external_metrics_provider.local_copy_refresh_rate")
-	return &datadogProvider{
+	d := &datadogProvider{
 		client: client,
 		mapper: mapper,
-		values: make(map[provider.CustomMetricInfo]int64),
 		store:  store,
 		maxAge: maxAge,
 	}
+	go d.externalMetricsSetter(ctx)
+	return d
+}
+
+type timer struct {
+	period time.Duration
+	timer  time.Timer
+}
+
+func (t *timer) resetTimer() {
+	t.timer.Stop()
+	t.timer = *time.NewTimer(t.period)
+}
+
+func createTimer(period time.Duration) *timer {
+	return &timer{period, *time.NewTimer(period)}
+}
+
+func (p *datadogProvider) externalMetricsSetter(ctx context.Context) {
+	log.Infof("Starting async loop to collect External Metrics")
+	tick := time.NewTicker(time.Duration(p.maxAge) * time.Second)
+	defer tick.Stop()
+	out := createTimer(3 * time.Duration(p.maxAge) * time.Second)
+	defer out.timer.Stop()
+
+	// If we exceed 3 retries trying to access the ConfigMap, we permafail and stop trying to refresh the External Metrics.
+	ctxCancel, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	for {
+		select {
+		case <-tick.C:
+			var externalMetricsList []externalMetric
+			// TODO as we implement a more resilient logic to access a potentially deleted CM, we should pass in ctxCancel in case of permafail.
+			rawMetrics, err := p.store.ListAllExternalMetricValues()
+			if err != nil {
+				log.Errorf("Could not list the external metrics in the store: %s", err.Error())
+				p.serving = false
+				break
+			}
+
+			for _, metric := range rawMetrics {
+				// Only metrics that exist in Datadog and available are eligible to be evaluated in the HPA process.
+				if !metric.Valid {
+					continue
+				}
+				var extMetric externalMetric
+				extMetric.info = provider.ExternalMetricInfo{
+					Metric: metric.MetricName,
+				}
+				// Avoid overflowing when trying to get a 10^3 precision
+				q, err := resource.ParseQuantity(fmt.Sprintf("%v", metric.Value))
+				if err != nil {
+					log.Errorf("Could not parse the metric value: %v into the exponential format", metric.Value)
+					continue
+				}
+				extMetric.value = external_metrics.ExternalMetricValue{
+					MetricName:   metric.MetricName,
+					MetricLabels: metric.Labels,
+					Value:        q,
+				}
+				externalMetricsList = append(externalMetricsList, extMetric)
+			}
+			p.externalMetrics = externalMetricsList
+			p.timestamp = metav1.Now().Unix()
+			p.serving = true
+
+			out.resetTimer()
+
+		case <-out.timer.C:
+			cancel()
+
+		case <-ctxCancel.Done():
+			p.serving = false
+			log.Error("Received instruction to Terminate collection of External Metrics, stopping async loop")
```

Comment:
Won't the main context be cancel when legitimately stopping the cluster-agent, hence should this be an `Error`?

Reply:
It will indeed be legitimate. Updating this to info, and adding an extra error log for the timeout case,.